### PR TITLE
Parallel ERK method

### PR DESCRIPTION
**Pull request by saurabhkgp21**
A 5-parallel 2-processor ERK formula from Kutta’s 3-parameter family of 5th-order 6-stage schemes.

@ChrisRackauckas Which test should be added for threading?

### CODE REVIEW

File: src/alg_utils.jl [hunk 1]
```diff
@@ -134,6 +134,7 @@ alg_order(alg::Exprb32) = 3
 alg_order(alg::Exprb43) = 4
 alg_order(alg::Anas5) = 5
 alg_order(alg::RK46NL) = 4
+alg_order(alg::PaRK2p5) = 5
```
Comment (ChrisRackauckas) [inline src/alg_utils.jl]: KuttaPRK2p5

File: src/perform_step/park_perform_step.jl [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  @unpack α21,α31,α32,α41,α42,α43,α5_6 = cache.tab
+  @unpack β1,β3,β5,β6,c2,c3,c4,c5_6 = cache.tab
+
+  f( k1, uprev, p, t)
+  f( k2, uprev + dt*α21*k1, p, t + c2*dt)
+  f( k3, uprev + dt*(α31*k1 + α32*k2), p, t + c3*dt)
+  f( k4, uprev + dt*(α41*k1 + α42*k2 + α43*k3), p, t + c4*dt)
+
+  if integrator.alg.threading == false
+    f( k5_6[1], uprev + dt*(α5_6[1,1]*k1 + α5_6[1,2]*k2 + α5_6[1,3]*k3 + α5_6[1,4]*k4),
```
Comment (ChrisRackauckas) [inline src/perform_step/park_perform_step.jl]: the `u` term here allocates a ton

File: src/perform_step/park_perform_step.jl [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+@muladd function perform_step!(integrator,cache::PaRK2p5Cache,repeat_step=false)
+  @unpack t,dt,uprev,u,f,p = integrator
+  @unpack k,k1,k2,k3,k4,k5_6,fsalfirst = cache
+  @unpack α21,α31,α32,α41,α42,α43,α5_6 = cache.tab
+  @unpack β1,β3,β5,β6,c2,c3,c4,c5_6 = cache.tab
+
+  f( k1, uprev, p, t)
+  f( k2, uprev + dt*α21*k1, p, t + c2*dt)
+  f( k3, uprev + dt*(α31*k1 + α32*k2), p, t + c3*dt)
+  f( k4, uprev + dt*(α41*k1 + α42*k2 + α43*k3), p, t + c4*dt)
```
Comment (ChrisRackauckas) [inline src/perform_step/park_perform_step.jl]: allocating, and also probably fails on the + between arrays

File: test/ode/ode_convergence_tests.jl [hunk 4]
```diff
@@ -181,6 +181,16 @@ for i = 1:2
   sim19 = test_convergence(dts,prob,Hairer4())
   @test sim19.𝒪est[:final] ≈ 4 atol=testTol
 
+  sim20 = test_convergence(dts,prob,PaRK2p5(threading=true))
```
Comment (ChrisRackauckas) [inline test/ode/ode_convergence_tests.jl]: put it with the ERK methods

File: src/perform_step/park_perform_step.jl [hunk 5]
```diff
@@ -0,0 +1,78 @@
+function initialize!(integrator,cache::PaRK2p5ConstantCache)
```
Comment (ChrisRackauckas) [inline src/perform_step/park_perform_step.jl]: just PRK for the file names

File: test/ode/ode_convergence_tests.jl (hunk 4 above)
Comment (saurabhkgp21) [reply test/ode/ode_convergence_tests.jl]: Should I also add complex tests?

Comment (ChrisRackauckas) [reply test/ode/ode_convergence_tests.jl]: I don't think that's needed here? There's already complex tests?

Comment (saurabhkgp21) [reply test/ode/ode_convergence_tests.jl]: https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/5390222e6dd85e8558c23e0cbed264d6e919e097/test/complex_tests.jl#L14
These explicit methods do have separate complex tests

File: src/perform_step/prk_perform_step.jl [hunk 6]
```diff
@@ -57,19 +57,26 @@ end
...
+    @. u = uprev + dt*(α5_6[1,1]*k1 + α5_6[1,2]*k2 + α5_6[1,3]*k3 + α5_6[1,4]*k4)
+    f( k5_6[1], u, p, t + c5_6[1]*dt)
+
+    @. u = uprev + dt*(α5_6[2,1]*k1 + α5_6[2,2]*k2 + α5_6[2,3]*k3 + α5_6[2,4]*k4)
+    f( k5_6[2], u, p, t + c5_6[2]*dt)
   else
     Threads.@threads for i in [1,2]
-      f( k5_6[i], uprev + dt*(α5_6[i,1]*k1 + α5_6[i,2]*k2 + α5_6[i,3]*k3 + α5_6[i,4]*k4),
-           p, t + c5_6[i]*dt)
+      @. u = uprev + dt*(α5_6[i,1]*k1 + α5_6[i,2]*k2 + α5_6[i,3]*k3 + α5_6[i,4]*k4)
```
Comment (ChrisRackauckas) [inline src/perform_step/prk_perform_step.jl]: That's not thread safe

Comment (saurabhkgp21) [reply src/perform_step/prk_perform_step.jl]: yeah.. sorry I somehow missed it
Should I pre-allocate separate memory for both threads or let it allocate memory on the run? While pre-allocating, same memory can also be used for calculation in parent thread.

Comment (ChrisRackauckas) [reply src/perform_step/prk_perform_step.jl]: you'll need to make use of `tmp` for one thread and you can use `u` for the other. Then you can make it `tmps = (u,tmp)`, and grab from `tmps` dependent on `i`.

**Pull request merged**